textdecoder

juj opened the pull request:
Use TextDecoder to decode UTF-8 and UTF-16 strings. Closes #4402.

Not quite yet to merge, needs the fallback path testing and a bit more benchmarking.

kripken commented on src/preamble.js:
```diff
@@ -515,39 +515,49 @@ function stringToAscii(str, outPtr) {
 // Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
 
+var utf8_decoder = new TextDecoder('utf8');
```
maybe `UTF8Decoder` for consistency?

kripken commented on src/preamble.js:
```diff
@@ -515,39 +515,49 @@ function stringToAscii(str, outPtr) {
+  var endPtr = idx;
+  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
+  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
+  while(u8Array[endPtr]) ++endPtr;
```
space after `while`

kripken commented on src/preamble.js:
```diff
@@ -669,17 +679,27 @@ function lengthBytesUTF8(str) {
+  var endPtr = ptr;
+  // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
+  // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
+  while({{{ makeGetValue('endPtr', 0, 'i16') }}}) endPtr += 2;
```
space after `while`

kripken commented on src/preamble.js:
```diff
@@ -669,17 +679,27 @@ function lengthBytesUTF8(str) {
+    var str = '';
+    while (1) {
+      var codeUnit = {{{ makeGetValue('ptr', 'i*2', 'i16') }}};
+      if (codeUnit == 0)
```
same line or `{}` (looks like this was an error already present)

juj commented:
Addressed review comments and added tests. Also added an option `-s TEXTDECODER=0/1` to allow disabling the feature if needed. This is mainly to allow easy performance testing for users in their environments, since it is difficult to offer conclusive testing in a synthetic test environment for UTF strings (e.g. some workloads might have mostly 7-bit ASCII strings, whereas others could be 4-byte UTF-8 heavy)

kripken commented:
Overall lgtm. The one concern I have are the new alignment requirements this adds. This might break existing code, although I guess very rarely?

juj commented:
Actually, the alignment requirements are old and have been present already earlier, see e.g. https://github.com/kripken/emscripten/blob/master/src/preamble.js#L677 that does a HEAP16 load. Only when updating this PR I realized we have this requirement, and added the asserts in for diligence. It would be possible to add support for unaligned cases, but probably not critical for now.

kripken commented:
I see, thanks. Yeah, since this was the case before, it all sounds good.
